### ISSUE TITLE
Revisit locking design for State object

### ISSUE THREAD

**Issue by karlfloersch** [plasma-group/plasma-operator#17]
**Is your feature request related to a problem? Please describe.**
The state object requires locking logic when starting a new block or ingesting transactions & deposits.  An example of a race condition without locking is as follows:

1) Transaction A & B are ingested at the same time.
2) Transaction A transfers range to Alice in block 5, transaction B transfers the same range to Bob in block 5. Both transactions are valid, but only one can be included as they have the same block number & transfer the same range twice.
3) Because there is no lock, both transactions make it past the validity checks concurrently.
4) Both transactions are applied to the state and saved in the `tx_log`... uh oh!
5) Consumers of the `tx_log` now have two transactions effecting the same range in the same block! Oh no--that's invalid!

I am currently solving this by locking the sending account when a transaction is added. That means that each of these txs would attempt to acquire a lock for the same sender & one would fail.

### How do locks currently get acquired and released?
Right now, locking is *very* primitive. Lock acquiring looks like this:

```
  attemptAcquireLocks (keywords) {
    keywords.push('all')
    if (keywords.some((val) => { return this.lock.hasOwnProperty(val) })) {
      // Failed to acquire locks
      return false
    }
    // Acquire locks
    for (let i = 0; i < keywords.length - 1; i++) {
      this.lock[keywords[i]] = true
    }
    return true
  }
```

Where `keywords` is every sender address in every transfer record of the transaction being added. If a lock fails to be acquired, then we wait some amount of time & try again with:

```
    while (!this.attemptAcquireLocks([type])) {
      // Wait before attempting again
      await timeout(timeoutAmt())
    }
```

### Problems with this approach
I used this as a quick and dirty solution but I believe it is a bit dangerous in that it may fill up the event queue if there are a large number of callbacks competing for the same lock. I don't actually know how bad this is but I imagine it could be a problem if someone were to intentionally fill up the event queue with tons of txs which are from the same sender.

### Solutions?
I don't know the best way to approach this I just intuit there's a solution I haven't yet explored. Maybe to do with queues and promises, but my thoughts in this area are not developed enough to describe at the moment.

I'd love feedback!

### Some resources I've taken a look at...
- https://medium.engineering/wait-dont-touch-that-a211832adc3a
- https://stackoverflow.com/questions/5346694/how-to-implement-a-lock-in-javascript
- https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred

**Comment by ben-chain** [comment]
For what it's worth, I will note that the race condition you describe does not produce an "invalid" block in the same way that would be used in an invalid history challenge.  At the leaf parsing step of the [merkle sum tree](https://plasma-core.readthedocs.io/en/latest/specs/sum-tree.html), all but one of the transactions would be given a sum of 0 and so those branches would not be able to be used in an exit game at all.  So, the race condition doesn't actually invalidate anything in the sense of having to exit or lose money.  

That said, it obviously seems like horrible practice to not solve this issue--especially if the operator is signing off on inclusion promise attestations to get the "0-conf"-like property.

Another note--locking based on account does have the issue that it restricts the user from sending more than one transaction per block.  Ideally, we would want to lock by range... not sure how much overhead this would end up adding.

It does seem like a proper queue module could be useful here depending on how quick locking/unlocking is--from [queue.js](http://code.iamkate.com/javascript/queues/), "it can be significantly faster than using arrays."  I'm not sure whether the DOS issue you described above is preventable without proper anti-DOS measures, though.

On recieving a transaction, you'd add it to the queue, with a callback to include/log the transaction, and then you would just repeatedly dequeue in a loop:
  1) check if any parts of the ranges are locked, if so, reject
  2) lock all the ranges,
  3) accept transaction (run callback)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 300.0 DAI (300.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $53,237.11 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by mvayngrib** [comment]
@karlfloersch how are you measuring the memory limit / speed issues?

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 4 weeks, 1 day from now.
Please review their action plans below:


**1) [sprimage](https://gitcoin.co/profile/sprimage) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?mutate_worker_action=approve&worker=sprimage) | [reject worker](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?mutate_worker_action=reject&worker=sprimage))_.**

Hi, I can use Promises to implement your lock idea for transactions. I'll be building it with efficiency in mind to ensure your system isn't slowed down. I have a lot of experience with Promises (i work with apis a lot and it will help). With Promise.resolve, we can prevent the tx mixup without any risk of overloading the system as each transaction will be resolved asynchronously and errors caught individually. I can get this done in less than a week too I'm sure. Looking forward to working with you

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250).

**Comment by mvayngrib** [comment]
is execution in order of invocation important? Currently, if the attempted lock acquisitions are something like 1. `['a', 'b']`, 2. `['b', 'c']`, 3. `['a', 'c']`, 4. `['a', 'b']`, 4 might end up executing in parallel with 2, and before 3

if execution in order of invocation is not important, then at every lock release, u can check the enqueued operations and choose the maximum number that can be run in parallel, as opposed to just the luck of the draw

**Comment by karlfloersch** [comment]
@mvayngrib 

> if execution in order of invocation is not important, then at every lock release, u can check the enqueued operations and choose the maximum number that can be run in parallel, as opposed to just the luck of the draw

This is exactly right. We want a datastructure which makes it easy to "choose the maximum number of parallelizable transactions" and then begin processing those transactions. The tricky part is managing this complex queue.

**Comment by mvayngrib** [comment]
@karlfloersch after playing with it on paper, and discussing with a friend, I'm no longer sure it matters whether parallelizable tasks run earlier or later. I think your locking design is actually fine efficiency-wise, the main issues being:
- it's embedded in state.js (lots of lock/unlock)
- it uses timeouts
- this kind of intersecting-lock-set queue might be useful on its own, in other parts of the project

i see @sprimage has been approved, but if he changes his mind, i'd like to give this a try

**Comment by Sprimage** [comment]
Yea @mvayngrib i figured the polling loop too was an issue and decided using https://www.npmjs.com/package/promise-queue should wrap it up. If you're really interested, we can work together to come up with the best solution as soon as possible

**Comment by Sprimage** [comment]
Ok so queuing promises can take a lot of memory. What if we move the job queue out of program memory and use something like ```fs```? So each file represents a job (the json of the functions args) and we loop through the dir to execute each job in the queue and delete the file once executed?

**Comment by mvayngrib** [comment]
@Sprimage fs sounds like a big compromise speed-wise. Blocks are processed one a time, what's the max # txs in a block? Also, as far as I understand, queueing is an exception rather than a rule. Txs can be processed in parallel and only queue up when they collide on a lock (as they are currently)

**Comment by karlfloersch** [comment]
@mvayngrib 
> what's the max # txs in a block?

This depends on the block time & tps as there is no hard upper limit on txs in a block. Right now the block time is set to 60 seconds, and I've benchmarked the state object alone and gotten around 3k tps on my laptop. So if we were to make the express API handle more tps, theoretically with 1 minute block times we could get blocks with 180,000 transactions, which would be approx 18MB.

**Comment by mvayngrib** [comment]
@karlfloersch are you considering using a persistent buffer of some sort, e.g. `redis` or `fs`, or is that out of the scope?

**Comment by Sprimage** [comment]
@mvayngrib

> Also, as far as I understand, queueing is an exception rather than a rule. Txs can be processed in parallel and only queue up when they collide on a lock (as they are currently)

I know queuing is an exception. But the same is the root of this issue isn't it?

**Comment by Sprimage** [comment]
Also external dependencies and platform specific dependencies would make ```fs``` the better option compared to redis. @karlfloersch do you think ```fs``` will work?

**Comment by Sprimage** [comment]
If you have slack or something similar so we can discuss this in real time and consider the solution together would be great too

**Comment by gitcoinbot** [comment]
@sprimage Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=1) | [3 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=3) | [5 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=5) | [10 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=10) | [100 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=100)

**Comment by gitcoinbot** [comment]
@sprimage Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=1) | [3 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=3) | [5 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=5) | [10 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=10) | [100 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=100)

**Comment by gitcoinbot** [comment]
@sprimage Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=1) | [3 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=3) | [5 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=5) | [10 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=10) | [100 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=100)

**Comment by gitcoinbot** [comment]
@sprimage Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=1) | [3 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=3) | [5 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=5) | [10 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=10) | [100 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=100)

**Comment by gitcoinbot** [comment]
@sprimage Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=1) | [3 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=3) | [5 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=5) | [10 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=10) | [100 days](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250?snooze=100)

**Comment by mvayngrib** [comment]
this still alive?

**Comment by ceresstation** [comment]
Hey @mvayngrib are you interested in taking this on? I think we should be passing it back to the crowd at this point.

**Comment by mvayngrib** [comment]
@ceresstation sure, i can give it a try. I already applied though gitcoin

**Comment by ceresstation** [comment]
Just accepted :)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 300.0 DAI (300.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@mvayngrib](https://gitcoin.co/profile/mvayngrib)



@ceresstation please take a look at the submitted work:
* [PR](https://github.com/plasma-group/plasma-chain-operator/pull/64) by @mvayngrib

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $40,322.07 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by mvayngrib** [comment]
poke @ceresstation

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 300.0 DAI (300.0 USD @ $1.0/DAI)  attached to this issue has been approved & issued to @mvayngrib.__  

 
* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/plasma-group/plasma-chain-operator/17/2250)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $48,523.51 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
